### PR TITLE
Fix typing of Lifespan to allow subclasses of Starlette

### PR DESCRIPTION
**Pull request by adriangb**
Fixes https://github.com/tiangolo/fastapi/discussions/9241#discussioncomment-5273143

### CODE REVIEW

File: tests/test_applications.py [hunk 1]
```diff
@@ -534,3 +534,14 @@ def get_app() -> ASGIApp:
...
+
+def test_lifespan_typing():
+    class App(Starlette):
+        pass
+
+    @asynccontextmanager
+    async def lifespan(app: App) -> AsyncIterator[None]:
+        yield
+
+    App(lifespan=lifespan)
```
Comment (adriangb) [inline tests/test_applications.py]: Not sure if we actually enforce mypy on our tests, but pylance reports no issues

**Review by tiangolo — APPROVED** [review]
Awesome, thank you @adriangb! This looks good.

I just tested this branch locally to confirm it works, and I made the corresponding PR in FastAPI here, I'll merge that and release once this is merged and released: https://github.com/tiangolo/fastapi/pull/9245

File: starlette/applications.py [hunk 2]
```diff
@@ -257,3 +257,6 @@ def decorator(func: typing.Callable) -> typing.Callable:
             return func
 
         return decorator
+
+
+AppType = typing.TypeVar("AppType", bound=Starlette)
```
Comment (Kludex) [inline starlette/applications.py]: Can we move this to the top, or is there an issue with string typing?

File: starlette/types.py [hunk 3]
```diff
@@ -1,7 +1,6 @@
 import typing
 
-if typing.TYPE_CHECKING:
-    from starlette.applications import Starlette
+T = typing.TypeVar("T")
```
Comment (Kludex) [inline starlette/types.py]:
```suggestion
AppType = typing.TypeVar("AppType")
```

File: tests/test_applications.py [hunk 4]
```diff
@@ -534,3 +534,14 @@ def get_app() -> ASGIApp:
...
+
+def test_lifespan_typing():
+    class App(Starlette):
+        pass
+
+    @asynccontextmanager
+    async def lifespan(app: App) -> AsyncIterator[None]:  # pragma: no cover
+        yield
+
+    App(lifespan=lifespan)
```
Comment (Kludex) [inline tests/test_applications.py:550]: Does this test fails before this PR?

**Review by Kludex — CHANGES_REQUESTED** [review]

**Comment by Kludex** [comment]
The lifespan parameter on `routing.py` doesn't get a parameter inside the `Lifespan`, is it intended? My IDE is more permissive than yours, so I don't have problems there, but knowing your configuration (:laughing:), your IDE is probably complaining over there.

**Comment by adriangb** [comment]
Good point. We probably have to add an Any

File: tests/test_applications.py (hunk 4 above)
Comment (adriangb) [reply tests/test_applications.py:550]: No, not really, I’m not sure how to force it to fail since we don’t run mypy on tests right?

Comment (Kludex) [reply tests/test_applications.py:550]: Maybe assert_type on the constructor? Idk

Comment (adriangb) [reply tests/test_applications.py:550]: Ok mypy does pick it up:

Before this change:
```
tests/test_applications.py:550: error: Argument "lifespan" to "App" has incompatible type "Callable[[App], _AsyncGeneratorContextManager[None]]"; expected "Union[Callable[[Starlette], AbstractAsyncContextManager[None]], Callable[[Starlette], AbstractAsyncContextManager[Mapping[str, Any]]], None]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

File: starlette/applications.py (hunk 2 above)
Comment (adriangb) [reply starlette/applications.py]: No we can't. It needs to reference `Starlette` so it has to come after it.

Comment (Kludex) [reply starlette/applications.py]: Just tested here, and it works before...

File: starlette/applications.py [hunk 5]
```diff
@@ -11,6 +11,8 @@
 from starlette.routing import BaseRoute, Router
 from starlette.types import ASGIApp, Lifespan, Receive, Scope, Send
 
+AppType = typing.TypeVar("AppType", bound="Starlette")
+
```
Comment (adriangb) [inline starlette/applications.py:15]: Does this work? If so 👍🏻

File: starlette/applications.py [hunk 6]
```diff
@@ -11,6 +11,8 @@
 from starlette.routing import BaseRoute, Router
 from starlette.types import ASGIApp, Lifespan, Receive, Scope, Send
 
+AppType = typing.TypeVar("AppType", bound="Starlette")
```
Comment (Kludex) [inline starlette/applications.py:14]: Check if you are fine with this, please. @adriangb

**Review by Kludex — APPROVED** [review]

File: starlette/applications.py (hunk 5 above)
Comment (Kludex) [reply starlette/applications.py:15]: yes

**Pull request merged**